streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -1012,6 +1022,8 @@ private void prepareChangelogs(final Map<TaskId, Task> tasks,
                 // no records to restore; in this case we just initialize the sensor to zero
                 final long recordsToRestore = Math.max(changelogMetadata.restoreEndOffset - startOffset, 0L);
                 task.recordRestoration(time, recordsToRestore, true);
+            }  else if (changelogMetadata.stateManager.taskType() == TaskType.STANDBY && storeMetadata.endOffset() != null) {
```

Comment:
Why do we only invoke the standby listener when we have the `endOffset` filled out? I don't know off the top of my head when it would be empty or not, but I would think the listener callbacks are still useful even without this one field filled in. Also, users might rely on the standby listener callbacks being invoked. I think we should guarantee that the listener is always called, at least in the absence of app-wide errors that cause a shutdown)

If you're wondering what to do in the case of it being null, I would suggest just passing in  `-1L` as a sentinel value. As long as we mention that this is a possibility in the javadocs for the `endOffset` argument, I don't see any problem with leaving it up to the user to decide how to react in this case

Reply:
That's a really good point, most (close to all) of the time we won't actually know the `endOffset` when `#onUpdateStart` is invoked. I think it's totally reasonable to remove this parameter from that callback. It's probably way more important/useful to track the end offset as it changes over the course of the standby updates anyways, compared to `#onRestoreStart` where the end offset never advances during restoration but it's nice to know it up front to compute how much there is to restore.

So I'm in agreement with 2), you can just send a quick update to the KIP thread with a brief explanation. That said, I think we should still try to guarantee that the listener callbacks are always invoked, even if we don't know the end offset. 
 In particular `#onUpdateSuspended`, which I imagine might be used to clean up resources (such as metrics) created during `#onUpdateStart`. One could argue that since `#onBatchLoaded` is only ever called after a successful poll and so the end offset should always be known in that case, but just for the sake of consistency, I think we should treat both callbacks the same and invoke them whether `endOffset` is known or not. Also, the end offset is ultimately just one piece of information and not the sole source of value, so it feels weird to skip over the listener altogether just because a single field is missing.

So I guess overall, I actually think we should do both of @coltmcnealy-lh 's suggestions, 1 and 2. As for whether to use `-1`, `null`, or `Optional.empty()`, I'll leave that up to the KIP authors. Of course since you should probably update the mailing list thread with this change as well, if you don't have a strong preference yourself, you can also just put the options out there and see what everyone agrees on.